File: src/main/java/com/murex/tbw/finance/TaxRule.java
package com.murex.tbw.finance;

import com.google.common.collect.ImmutableMap;
import com.murex.tbw.domain.book.Book;
import com.murex.tbw.domain.book.Novel;
import com.murex.tbw.domain.country.Country;
import com.murex.tbw.domain.country.Language;

import java.util.Map;

public final class TaxRule {

    private static final Map<String, Double> TAX_RATES = ImmutableMap.<String, Double>builder()
            .put("USA", 1.15d)
            .put("France", 1.25d)
            .put("UK", 1.20d)
            .put("Spain", 1.10d)
            .put("China", 1.35d)
            .put("Japan", 1.30d)
            .put("Australia", 1.13d)
            .put("Germany", 1.22d)
            .build();

    public static double getApplicableRate(Country invoiceCountry, Book book) {
        if (invoiceCountry.getName().equals("Germany")) {
            if (book.getAuthor().getNationality().getName().equals("Germany")) {
                return 1.05;
            }
        }
        if (invoiceCountry.getName().equals("USA")) {
            if (book.getClass().equals(Novel.class)) {
                return getTaxRate(invoiceCountry.getName()) * 0.98;
            }
        }
        if (invoiceCountry.getName().equals("UK")) {
            if (book.getClass().equals(Novel.class)) {
                return getTaxRate(invoiceCountry.getName()) * 0.93;
            }
        }
        if (invoiceCountry.getName().equals("China")) {
            if (!book.getLanguage().equals(Language.MANDARIN)) {
                return 1d;
            }
        }
        if (invoiceCountry.getName().equals("Spain")) {
            if (!book.getLanguage().equals(Language.SPANISH)) {
                return 1d;
            }
        }
        return getTaxRate(invoiceCountry.getName());
    }

    private static double getTaxRate(String country) {
        return TAX_RATES.get(country);
    }
}

File: typescript/src/com/murex/tbw/finance/TaxRule.ts
import { Book } from "../domain/book/Book";
import { Country } from "../domain/country/Country";
import * as Immutable from "immutable";
import { Language } from "../domain/country/Language";
import { isNovel } from "../domain/book/Novel";

const taxRates = Immutable.Map({
  USA: 1.15,
  France: 1.25,
  UK: 1.2,
  Spain: 1.1,
  China: 1.35,
  Japan: 1.3,
  Australia: 1.13,
  Germany: 1.22,
});

function getTaxRate(countryName: string): number {
  const newLocal = taxRates.get(countryName);
  if (newLocal === undefined) {
    throw new Error("Unhandled country name: " + countryName);
  }
  return newLocal;
}

export function getApplicableRate(invoiceCountry: Country, book: Book): number {
  if (invoiceCountry.name === "Germany") {
    if (book.author.nationality.name === "Germany") {
      return 1.05;
    }
  }
  if (invoiceCountry.name === "USA") {
    if (isNovel(book)) {
      return getTaxRate(invoiceCountry.name) * 0.98;
    }
  }
  if (invoiceCountry.name === "UK") {
    if (isNovel(book)) {
      return getTaxRate(invoiceCountry.name) * 0.93;
    }
  }
  if (invoiceCountry.name === "China") {
    if (book.language !== Language.Mandarin) {
      return 1.0;
    }
  }
  if (invoiceCountry.name === "Spain") {
    if (book.language !== Language.Spanish) {
      return 1.0;
    }
  }
  return getTaxRate(invoiceCountry.name);
}
